File: solution/implementation/21608/main.py
# Authored by : gusdn3477
# Co-authored by : -
# Link : http://boj.kr/178ebc4de06c4c1fb718d3038ec42d50

import sys

def input():
    return sys.stdin.readline().rstrip()

N = int(input())
total = 0
nx = [-1, 0, 1, 0]
ny = [0, -1, 0, 1]
stu = []
feel = [[0 for i in range(N)] for j in range(N)]
arr = [[0 for i in range(N)] for j in range(N)]
save = [[] for i in range(N**2+1)]
for i in range(N**2):
    a = list(map(int, input().split()))
    stu.append(a)
    save[a[0]].append(a[1:])
for i in range(N**2):
    temp = []
    for x in range(N):
        for y in range(N):
            if arr[x][y] != 0:
                continue
            like = 0
            empty = 0
            for z in range(4):
                dx = x + nx[z]
                dy = y + ny[z]
                if dx < 0 or dx >= N or dy < 0 or dy >= N:
                    continue
                if  arr[dx][dy] in stu[i][1:]:
                    like += 1
                if arr[dx][dy] == 0:
                    empty += 1
            temp.append((like, empty, (x,y)))
    temp.sort(key = lambda x : (-x[0], -x[1], x[2]))
    arr[temp[0][2][0]][temp[0][2][1]] = stu[i][0]
for i in range(N):
    for j in range(N):
        now = arr[i][j]
        near = 0
        for z in range(4):
            dx = i + nx[z]
            dy = j + ny[z]
            if dx < 0 or dx >= N or dy < 0 or dy >= N:
                continue
            if arr[dx][dy] in save[now][0]:
                near += 1
        feel[i][j] = near
for i in range(N):
    for j in range(N):
        if feel[i][j] == 1:
            total += 1
        elif feel[i][j] == 2:
            total += 10
        elif feel[i][j] == 3:
            total += 100
        elif feel[i][j] == 4:
            total += 1000
print(total)

File: solution/implementation/21608/main.js
// Authored by : highjoon
// Co-authored by : -
// Link : http://boj.kr/dfbb944e04d3425b92c43bc40e9d5446

let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

function solution(input) {
    let N = Number(input[0]);
    input = input.slice(1).map((v) => v.split(" ").map(Number));
    let board = Array.from({ length: N }, () => Array(N).fill(0));
    const dx = [1, 0, -1, 0];
    const dy = [0, 1, 0, -1];
    let favorites = {};
    let students = [];
    let answer = 0;

    for (let x of input) {
        students.push(x[0]);
        favorites[x[0]] = x.slice(1);
    }

    for (let student of students) {
        let candidates = {};
        for (let row = 0; row < N; row++) {
            for (let col = 0; col < N; col++) {
                if (board[row][col] !== 0) continue;
                let count = [0, 0]; // 우선 순위 수, 비어있는 칸 수,
                for (let i = 0; i < dx.length; i++) {
                    const [nRow, nCol] = [row + dx[i], col + dy[i]];
                    if (nRow < 0 || nRow >= N || nCol < 0 || nCol >= N) continue;
                    if (favorites[student].includes(board[nRow][nCol])) count[0]++;
                    if (!board[nRow][nCol]) count[1]++;
                }
                const key = JSON.stringify(count);
                if (!candidates[key]) candidates[key] = [[row, col]];
                else candidates[key].push([row, col]);
            }
        }
        const sorted = Object.keys(candidates).sort((a, b) => {
            const [a0, a1] = JSON.parse(a);
            const [b0, b1] = JSON.parse(b);
            if (a0 !== b0) return b0 - a0;
            return b1 - a1;
        });
        const [newRow, newCol] = candidates[sorted[0]][0];
        board[newRow][newCol] = student;
    }

    for (let student of students) {
        let favCount = 0;
        for (let row = 0; row < N; row++) {
            for (let col = 0; col < N; col++) {
                if (board[row][col] === student) {
                    for (let i = 0; i < dx.length; i++) {
                        const [nRow, nCol] = [row + dx[i], col + dy[i]];
                        if (nRow < 0 || nRow >= N || nCol < 0 || nCol >= N) continue;
                        if (favorites[student].includes(board[nRow][nCol])) favCount++;
                    }
                }
            }
        }
        switch (favCount) {
            case 0:
                answer += 0;
                break;
            case 1:
                answer += 1;
                break;
            case 2:
                answer += 10;
                break;
            case 3:
                answer += 100;
                break;
            case 4:
                answer += 1000;
                break;
        }
    }

    return answer;
}

console.log(solution(input));
